### PR TITLE
Make sure TensorFlow also works on offline machines

### PR DESCRIPTION
**Pull request by laraPPr**
Depends on the new hook in this pr 
- https://github.com/EESSI/test-suite/pull/279

### CODE REVIEW

File: eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py [hunk 1]
```diff
@@ -1,9 +1,39 @@
+import os
+
 import tensorflow as tf
 import numpy as np
 
+if os.environ.get('EESSI_TEST_SUITE_NO_DOWNLOAD') == 'True':
```
Comment (laraPPr) [inline eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:6]: This is set in the configuration file under partition

Comment (laraPPr) [reply eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:6]: See https://github.com/laraPPr/EESSI-test-suite/blob/ed6ec51f75ac66d37c8c7851c34995ae24a7d02c/config/bsc_marenostrum5.py#L26-L30

File: eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py [hunk 2]
```diff
@@ -1,9 +1,39 @@
...
+else:
+    eessi_test_suite_download = True
+
 
 def mnist_dataset(batch_size, test_batch_size):
-    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
+    if eessi_test_suite_download:
+        (x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
+    else:
+        if "EESSI_TEST_SUITE_DOWNLOAD_DIR" in os.environ:
```
Comment (casparvl) [inline eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:16]: This is new, right? If so, it's use should be documented somewhere.

Comment (laraPPr) [reply eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:16]: Yes, it worked by setting it up in the configuration file of the system. I was waiting on writing the docs until you said that this is a good way to work arround no internet access.

File: eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py [hunk 3]
```diff
@@ -1,9 +1,39 @@
...
+                    x_train, y_train = data['x_train'], data['y_train']
+                    x_test, y_test = data['x_test'], data['y_test']
+            else:
+                msg = f'could not find {mnist_path} and cannot download.'
+                raise ValueError(msg)
+        else:
+            msg = 'The TensorFlow test requires EESSI_TEST_SUITE_DOWNLOAD_DIR to be set.'
+            msg += 'Since the dataset to run TensorFlow cannot be downloaded.'
+            raise ValueError(msg)
+    if eessi_test_suite_download:
```
[the remark this answers is not in the document]
Comment (casparvl) [reply eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:30]: We discussed, this section is indeed duplicate, the first one is the final one

File: eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py (hunk 2 above)
Comment (laraPPr) [reply eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:16]: Maybe try to use `resourcesdir` instead of having to set the environmentvariable in the configuration file

Comment (laraPPr) [reply eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:16]: @casparvl @smoors @satishskamath have any of you used the resourcesdir because I can't find any documentation other how you set the path in the configuration file

Comment (casparvl) [reply eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:16]: [Here](https://github.com/eth-cscs/cscs-reframe-tests/blob/85f208c3f4b0fd448696f068d482040fda431337/checks/apps/namd/namd_check.py#L51) is an example. It seems like you can get the current system from `self.curren_system`, and then it's a property of that.

Comment (casparvl) [reply eessi/testsuite/tests/apps/tensorflow/src/mnist_setup.py:16]: I wonder if we should somehow tag tests consistently if they require (pre-)downloaded data. The alternative to pre-downloading the data would be to just exclude tests with that label on offline systems.

I think it would be good to tag, but... with what? `requires_download`?

**Comment by smoors** [comment]
@laraPPr why not first download the dataset in a `local` test, i.e. not on the compute nodes but on the node where reframe is launched?
the only requirement is that this node has internet access, which is usually the case, as most people will launch reframe on a login node.
an added benefit is that the dataset is download only once for all the concrete test cases.

running a test locally should be doable with the `local` parameter, see:
https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#reframe.core.pipeline.RegressionTest.local

**Comment by laraPPr** [comment]
Mare Nostrum does not have login nodes with internet access. Which is a real pita.

**Comment by laraPPr** [comment]
And tensorflow redowloads everytime even when it is cashed.

File: eessi/testsuite/tests/apps/tensorflow/tensorflow.py [hunk 4]
```diff
@@ -97,3 +98,20 @@ def set_thread_count_args(self):
...
+        if self.current_partition.extras['internet_access'] == 'offline':
+            resourcesdir = self.current_system.resourcesdir
+            data = os.path.join(resourcesdir, self.module_name, 'datasets/mnist.npz')
+            if os.path.exists(data):
+                self.env_vars['EESSI_TEST_SUITE_DISABLE_DOWNLOAD'] = 'True'
+                self.env_vars['RFM_TENSORFLOW_DATA'] = data
+            else:
+                msg = f'{self.current_partition.name} does not have internet access. '
+                msg += f'And could not find {data}. '
+                msg += f'You can download the file running tf.keras.datasets.mnist.load_data() with {self.module_name} on a system with internet access.'
```
Comment (laraPPr) [inline eessi/testsuite/tests/apps/tensorflow/tensorflow.py:116]:
```suggestion
                msg += f'You can download the file running tf.keras.datasets.mnist.load_data() '
                msg += f'with {self.module_name} on a system with internet access.'
```

File: eessi/testsuite/tests/apps/tensorflow/tensorflow.py [hunk 5]
```diff
@@ -97,3 +98,21 @@ def set_thread_count_args(self):
...
+        if self.current_partition.extras['internet_access'] == 'offline':
+            resourcesdir = self.current_system.resourcesdir
+            data = os.path.join(resourcesdir, self.module_name, 'datasets/mnist.npz')
+            if os.path.exists(data):
+                self.env_vars['EESSI_TEST_SUITE_DISABLE_DOWNLOAD'] = 'True'
+                self.env_vars['RFM_TENSORFLOW_DATA'] = data
+            else:
+                msg = f'{self.current_partition.name} does not have internet access. '
+                msg += f'And could not find {data}. '
+                msg += f'You can download the file running tf.keras.datasets.mnist.load_data() '
```
Comment (laraPPr) [inline eessi/testsuite/tests/apps/tensorflow/tensorflow.py:116]:
```suggestion
                msg += 'You can download the file running tf.keras.datasets.mnist.load_data() '
```

**Comment by laraPPr** [comment]
I think we need to update that CI because it is not testing the TensorFlow tests not sure why `Tests for EESSI test suite, using EESSI production repo / test_with_eessi_pilot (2023.06) `

**Comment by laraPPr** [comment]
Converted to draft because I've now switched the other prs to use features instead of extras

**Review by casparvl — CHANGES_REQUESTED** [review]
I added 

```
FEATURES.OFFLINE,
```

to our local config. Then, the test got skipped. From the logs:

```
[2025-10-23T21:45:07] debug: reframe: [check_files_for_offline_run]: Warning: will exclude TensorFlow/2.13.0-foss-2023a tests on offline partitions.
Because reframe could not find ./TensorFlow/2.13.0-foss-2023a/datasets/mnist.npz.
You can download the file running tf.keras.datasets.mnist.load_data() with TensorFlow/2.13.0-foss-2023a on a system with internet access.
```

Then, I downloaded the dataset. Since I don't set a `resourcesdir` explicitly in my ReFrame config, it's using the current directory as default. Unfortunately, in `load_data(path=...)` the path is _relative_ to the `.keras/datasets` folder according to the documentation https://www.tensorflow.org/api_docs/python/tf/keras/datasets/mnist/load_data#args. Fortunately, a full path _is_ respected, so I could download it using:

```
mkdir -p TensorFlow/2.13.0-foss-2023a/datasets
python -c "import tensorflow as tf; tf.keras.datasets.mnist.load_data(path='/home/casparl/EESSI/test-suite/TensorFlow/2.13.0-foss-2023a/datasets/mnist.npz')"
``` 

Note the `mkdir -p`: the directory needs to exist, otherwise you get a `FileNotFoundError`.

I think we should improve the instruction so that it prints these _exact_ commands. ReFrame _knows_ what it's own resourcesdir is, we _can_ print that as part of the path.

Now the bad news: rerunning still gives me:

```
debug: reframe: [check_files_for_offline_run]: Warning: will exclude TensorFlow/2.13.0-foss-2023a tests on offline partitions.
Because reframe could not find ./TensorFlow/2.13.0-foss-2023a/datasets/mnist.npz.
You can download the file running tf.keras.datasets.mnist.load_data() with TensorFlow/2.13.0-foss-2023a on a system with internet access.
```
Reason is that the current working dir for a test is the location of the test file:

```
[2025-10-23T22:13:16] debug: reframe: [check_files_for_offline_run]: Current working dir: /gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow
[2025-10-23T22:13:16] debug: reframe: [check_files_for_offline_run]: Warning: will exclude TensorFlow/2.13.0-foss-2023a tests on offline partitions.
```

That's another reason why it's useful if it gets printed from the test itself: if the `resourcesdir` is `.`, it can use `os.getcwd()` instead.

**Comment by casparvl** [comment]
I'm also not sure why I get this redefinition issue in flake8. The line number seems wrong, and I really don't see a redefinition?

**Comment by casparvl** [comment]
> note that i had to use test.__class__.__name__, because test.unique_name was too unique :), i.e. it would still log every test instance.

Oh good point, forgot about that... But I'd say indeed `test.__class__.__name__` is the level of deduplication between log messages we were looking for.

**Review by casparvl — APPROVED** [review]
TLDR: approved!

Tested that the warning only gets printed once:
```
(eessi_test_venv) [casparl@int6 test-suite]$ reframe -c eessi/testsuite/tests/apps/tensorflow/ --run --system snellius:rome

[ReFrame Setup]
  version:           4.7.0-dev.3+20ccf7af
  command:           '/gpfs/home4/casparl/EESSI/eessi_test_venv/bin/reframe -c eessi/testsuite/tests/apps/tensorflow/ --run --system snellius:rome'
  launched by:       casparl@int6.local.snellius.surf.nl
  working directory: '/gpfs/home4/casparl/EESSI/test-suite'
  settings files:    '<builtin>', '/gpfs/home4/casparl/EESSI/test-suite/config/surf_snellius.py'
  check search path: (R) '/gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow'
  stage directory:   '/scratch-shared/casparl/reframe_output/staging'
  output directory:  '/gpfs/home4/casparl/EESSI/reframe_runs/output'
  log files:         '/gpfs/home4/casparl/EESSI/reframe_runs/logs/reframe_20251106_133851.log'

== Temporary log file in case of crash /scratch-local/45397/eb-8joat2kq/easybuild-4w0hya03.log
WARNING: Some partitions are labeled offline and will be excluded from EESSI_TensorFlow,
because ReFrame could not find the data at /gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow/TensorFlow/2.13.0-foss-2023a/datasets/mnist.npz.
To pre-download the data, please run (on a system with internet access):
mkdir -p /gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow/TensorFlow/2.13.0-foss-2023a/datasets
module load TensorFlow/2.13.0-foss-2023a
python -c "import tensorflow as tf; tf.keras.datasets.mnist.load_data('/gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow/TensorFlow/2.13.0-foss-2023a/datasets/mnist.npz')"
[==========] Running 0 check(s)
[==========] Started on Thu Nov  6 13:38:53 2025+0100

[----------] start processing checks
[----------] all spawned checks have finished

[  PASSED  ] Ran 0/0 test case(s) from 0 check(s) (0 failure(s), 0 skipped, 0 aborted)
[==========] Finished on Thu Nov  6 13:38:53 2025+0100
Log file(s) saved in '/gpfs/home4/casparl/EESSI/reframe_runs/logs/reframe_20251106_133851.log'
```

Tested the download instructions and then rerunning:

```
$ mkdir -p /gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow/TensorFlow/2.13.0-foss-2023a/datasets
module load TensorFlow/2.13.0-foss-2023a
python -c "import tensorflow as tf; tf.keras.datasets.mnist.load_data('/gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow/TensorFlow/2.13.0-foss-2023a/datasets/mnist.npz')"
Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/mnist.npz
11490434/11490434 [==============================] - 1s 0us/step
$ module purge
$ module load EESSI
Module for EESSI/2023.06 loaded successfully
$ reframe -c eessi/testsuite/tests/apps/tensorflow/ --run --system snellius:rome -t 1_node

[ReFrame Setup]
  version:           4.7.0-dev.3+20ccf7af
  command:           '/gpfs/home4/casparl/EESSI/eessi_test_venv/bin/reframe -c eessi/testsuite/tests/apps/tensorflow/ --run --system snellius:rome -t 1_node'
  launched by:       casparl@int6.local.snellius.surf.nl
  working directory: '/gpfs/home4/casparl/EESSI/test-suite'
  settings files:    '<builtin>', '/gpfs/home4/casparl/EESSI/test-suite/config/surf_snellius.py'
  check search path: (R) '/gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow'
  stage directory:   '/scratch-shared/casparl/reframe_output/staging'
  output directory:  '/gpfs/home4/casparl/EESSI/reframe_runs/output'
  log files:         '/gpfs/home4/casparl/EESSI/reframe_runs/logs/reframe_20251106_133701.log'

== Temporary log file in case of crash /scratch-local/45397/eb-9sbjju2p/easybuild-rkfbtw2u.log
[==========] Running 1 check(s)
[==========] Started on Thu Nov  6 13:37:04 2025+0100

[----------] start processing checks
[ RUN      ] EESSI_TensorFlow %scale=1_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /9864d0f5 @snellius:rome+default
[       OK ] (1/1) EESSI_TensorFlow %scale=1_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /9864d0f5 @snellius:rome+default
P: perf: 60312.18172813244 img/s (r:0, l:None, u:None)
[----------] all spawned checks have finished

[  PASSED  ] Ran 1/1 test case(s) from 1 check(s) (0 failure(s), 0 skipped, 0 aborted)
[==========] Finished on Thu Nov  6 13:37:55 2025+0100
Log file(s) saved in '/gpfs/home4/casparl/EESSI/reframe_runs/logs/reframe_20251106_133701.log'
```

Perfect!
